### PR TITLE
Update mynewt port of mcumgr and various fixes

### PR DESCRIPTION
**Pull request by vrahane**
- Updating mynewt port of MCUmgr
- Adding Stats, Logs, datetime and other default  commands support to MCUmgr
- Changing `mgmt_group_register()` in MCUmgr so that it allows re-registration of command groups
- Please refer https://github.com/apache/mynewt-core/issues/1871 for more context
- Fix dependency and build issues
- Adding support for log version 3 and image hash changes
- Adding support for image management lazy erase
- Adding support for config mgmt
- Updating cborattr to have similar support as mynewt-core
- Get all features in sync with mynewt-core
- Fix unit tests and add some more for cborattr
- Change log_mgmt to use fragmenting so that an entry can be divided into smaller chunks while reading.
- Fix mynewt port of os_mgmt
- Add changes for backwards compatibility
- Add sample app: `omp_svr` for trying out OICMGR over BLE/Serial
- Fix reboot_log for soft resets
Please refer https://github.com/apache/mynewt-core/pull/1917 for corresponding mynewt-core changes.

### CODE REVIEW

File: cmd/stat_mgmt/pkg.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+pkg.name: cmd/stat_mgmt
+pkg.description: 'OS management command handlers for mcumgr.'
```
Comment (mkiiskila) [inline cmd/stat_mgmt/pkg.yml]: Description should be updated.

File: cmd/stat_mgmt/port/mynewt/src/mynewt_stat_mgmt.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ * under the License.
+ */
+
+
+#include "stats/stats.h"
+#include "mgmt/mgmt.h"
+#include "sysinit/sysinit.h"
+#include "stat_mgmt/stat_mgmt.h"
+#include "stat_mgmt/stat_mgmt_impl.h"
+#include "../../../src/stat_mgmt_config.h"
```
Comment (mkiiskila) [inline cmd/stat_mgmt/port/mynewt/src/mynewt_stat_mgmt.c]: stat_mgmt_config.h is part of exported API? In that case it should be moved to that package's include directory.

File: util/pkg.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+pkg.name: util 
+pkg.description: Image management utility library
```
Comment (mkiiskila) [inline util/pkg.yml]: This is probably not how to classify the contents of this library.  Maybe Integer to string conversion?

File: cmd/img_mgmt/include/img_mgmt/img_mgmt.h [hunk 4]
```diff
@@ -21,12 +21,38 @@
...
+#include "../../src/img_mgmt_config.h"
+#include "mgmt/mgmt.h"
+
 struct image_version;
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+#define IMG_MGMT_MAX_IMGR         2
```
[the remark this answers is not in the document]
Comment (mkiiskila) [reply cmd/img_mgmt/include/img_mgmt/img_mgmt.h]: Actually, is this define used anywhere?

File: util/pkg.yml (hunk 3 above)
Comment (vrahane) [reply util/pkg.yml]: done

File: cmd/stat_mgmt/port/mynewt/src/mynewt_stat_mgmt.c (hunk 2 above)
Comment (vrahane) [reply cmd/stat_mgmt/port/mynewt/src/mynewt_stat_mgmt.c]: done and also moved other configs as well

File: cmd/stat_mgmt/pkg.yml (hunk 1 above)
Comment (vrahane) [reply cmd/stat_mgmt/pkg.yml]: updated

File: cmd/log_mgmt/src/log_mgmt.c [hunk 5]
```diff
@@ -63,36 +65,74 @@ static int
...
+     * Write entry data as byte string. Since this may not fit into single
+     * chunk of data we will write as indefinite-length byte string which is
+     * basically a indefinite-length container with definite-length strings
+     * inside.
+     */
+    err |= cbor_encoder_create_indef_byte_string(&rsp, &str_encoder);
+    for (off = 0; off < entry->len && !err; ) {
+        err |= cbor_encode_byte_string(&str_encoder, entry->data, entry->len);
+        off += entry->len;
+    }
```
[the remark this answers is not in the document]
Comment (vrahane) [reply cmd/log_mgmt/src/log_mgmt.c]: This code has changed now, I found another issue which was lingering. so, I am going to mark it as done.

File: cmd/log_mgmt/port/mynewt/src/mynewt_log_mgmt.c [hunk 6]
```diff
@@ -127,32 +124,54 @@ mynewt_log_mgmt_walk_cb(struct log *log, struct log_offset *log_offset,
...
-    entry.ts = ueh.ue_ts;
-    entry.index = ueh.ue_index;
-    entry.module = ueh.ue_module;
-    entry.level = ueh.ue_level;
-    entry.len = rc;
-    entry.data = mynewt_log_mgmt_walk_arg->body;
+    entry.len = entrylen;
+    entry.data = mynewt_log_mgmt_walk_arg->chunk;
+    rc = mynewt_log_mgmt_walk_arg->cb(&entry, mynewt_log_mgmt_walk_arg->arg);
+    if (rc) {
```
[the remark this answers is not in the document]
Comment (vrahane) [reply cmd/log_mgmt/port/mynewt/src/mynewt_log_mgmt.c]: done

File: cmd/img_mgmt/port/mynewt/src/mynewt_img_mgmt.c [hunk 7]
```diff
@@ -157,27 +392,124 @@ img_mgmt_impl_read(int slot, unsigned int offset, void *dst,
...
+{
+    int rc = 0;
+    struct flash_area sector;
+    const struct flash_area *cfa;
+
+    rc = flash_area_open(FLASH_AREA_IMAGE_1, &cfa);
+    if (rc != 0) {
+        return MGMT_ERR_EUNKNOWN;
+    }
+
```
[the remark this answers is not in the document]
Comment (vrahane) [reply cmd/img_mgmt/port/mynewt/src/mynewt_img_mgmt.c]: done

File: cmd/img_mgmt/pkg.yml [hunk 8]
```diff
@@ -24,7 +24,13 @@ pkg.homepage: "http://mynewt.apache.org/"
...
+    - '@apache-mynewt-mcumgr/cmd/img_mgmt/port/mynewt'
     - '@apache-mynewt-core/boot/split'
     - '@apache-mynewt-core/encoding/base64'
     - '@apache-mynewt-core/sys/flash_map'
-    - '@mynewt-mcumgr/mgmt'
+    - '@apache-mynewt-mcumgr/mgmt'
+    - '@apache-mynewt-mcumgr/cborattr'
+    - '@apache-mynewt-mcumgr/util'
+
+IMG_MGMT_LAZY_ERASE.cflags:
```
[the remark this answers is not in the document]
Comment (vrahane) [reply cmd/img_mgmt/pkg.yml]: you are right its not needed, I removed it.

File: cmd/img_mgmt/include/img_mgmt/img_mgmt_impl.h [hunk 9]
```diff
@@ -100,6 +104,71 @@ int img_mgmt_impl_write_image_data(unsigned int offset, const void *data,
...
+
+/**
+ * Erases image data at given offset
+ *
+ * @param offset                The offset within slot 1 to erase at.
+ * @param num_bytes             The number of bytes to erase.
+ * @return                      0 on success, MGMT_ERR_[...] code on failure.
+ */
+int
+img_mgmt_impl_erase_image_data(unsigned int off, unsigned int num_bytes);
```
[the remark this answers is not in the document]
Comment (vrahane) [reply cmd/img_mgmt/include/img_mgmt/img_mgmt_impl.h]: Changed this one and others as well.

File: cmd/img_mgmt/include/img_mgmt/img_mgmt.h [hunk 10]
```diff
@@ -44,11 +70,209 @@ extern "C" {
...
+ */
+int
+img_mgmt_state_confirm(void);
+
+/** @brief Generic callback function for events */
+typedef void (*img_mgmt_dfu_cb)(void);
+
+/** Callback function pointers */
+typedef struct
+{
```
[the remark this answers is not in the document]
Comment (vrahane) [reply cmd/img_mgmt/include/img_mgmt/img_mgmt.h]: done

File: cmd/img_mgmt/include/img_mgmt/img_mgmt.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (vrahane) [reply cmd/img_mgmt/include/img_mgmt/img_mgmt.h]: You are right, it is not used, I wonder why it was there in the first place but I removed it anyways.

File: cmd/img_mgmt/include/img_mgmt/img_mgmt.h [hunk 11]
```diff
@@ -21,12 +21,38 @@
 #define H_IMG_MGMT_
 
 #include <inttypes.h>
+#include "../../src/img_mgmt_config.h"
```
[the remark this answers is not in the document]
Comment (vrahane) [reply cmd/img_mgmt/include/img_mgmt/img_mgmt.h]: Done and moved others as well.

File: cborattr/src/cborattr.c [hunk 12]
```diff
@@ -421,4 +437,181 @@ cbor_read_mbuf_attrs(struct os_mbuf *m, uint16_t off, uint16_t len,
...
+{
+    int len;
+    int rc;
+
+    if (attr->omit) {
+        return 0;
+    }
+
+    len = strlen(attr->attribute);
+    rc = cbor_encode_text_string(enc, attr->attribute, len);
```
[the remark this answers is not in the document]
Comment (vrahane) [reply cborattr/src/cborattr.c]: done

**Comment by vrahane** [comment]
@mkiiskila I made changes as per your review comments, thanks.

**Comment by nvlsianpu** [comment]
@de-nordic Pleas review

**Comment by carlescufi** [comment]
This completely broke the Zephyr port @vrahane and @mkiiskila. We did not have time to react quickly enough, but I think we should strive for changes that are posted as a PR here to be tested on *both* Mynewt and Zephyr, since this is a cross-RTOS open source project.
Can we have a talk about this and avoid this situation again?

**Comment by vrahane** [comment]
@carlescufi Sure, we can have a chat about it. The work is not yet over, we will be making more changes to get the zephyr side working. @mlaz is working on getting the samples to work.

**Comment by carlescufi** [comment]
> @carlescufi Sure, we can have a chat about it. The work is not yet over, we will be making more changes to get the zephyr side working. @mlaz is working on getting the samples to work.

Thanks @vrahane. I think we need to sync up because @demo-nordic is also working on the same changes.

**Comment by de-nordic** [comment]
Pasting my comment from #mcumgr channel here:
basically it is no longer possible to compile/link mcumgr library within zephyr due to changes in mcumgr/cmd/img_mgmt/img_mgmt.c . There has been some impactful changes to this common code, that have been backed by changes in port/mynewt but at the same time port/zephyr has not been altered to provide missing functions. For example: img_mgmt_upload relies on img_mgmt_impl_upload_inspect which is now missing from port/zephyr implementation.

**Comment by de-nordic** [comment]
Right now I am at the stage when code compiles but does not link. I have paused in working on implementing missing functions and now I wait to sync up with you to avoid redundancy in work effort.

**Pull request merged**